Integrate the DNS search domain fallbacks

FR4NK-W opened the pull request:
Make use of the DNS search domain fallbacks:
If no DNS search domain is discovered by the other hinters before the timeout is about to expire, use the public IP(s) of the host and reverse DNS to discover candidate DNS search domains. If there are no public IPs, use external DNS servers to discover publicly reachable IPs.

FR4NK-W committed:
Integrate the DNS search domain fallbacks

Make use of the DNS search domain fallbacks:
If no DNS search domain is discovered by the other hinters before the timeout is about to expire,
use the public IP(s) of the host and reverse DNS to discover candidate DNS search domains.
If there are no public IPs, use external DNS servers to discover publicly reachable IPs.

marcfrei reviewed (approved):
LGTM

marcfrei commented:
Thanks, François.

As for exposing the external IP to third parties, I think we should shortly explain this in the README and maybe optionally also provide a command-line option to opt-out of this behavior.

I wasn't aware of a related issue in split DNS setups, thanks for pointing the problem out. Your proposed mitigation sounds reasonably easy to implement. Maybe we should thus do this, just to be extra careful?

FR4NK-W committed:
Add information to README about DSD fallback

Add information about DNS Search Domain fallback, use of the whoami service and public DNS resolver fallback.
Provide information how to avoid needing those fallbacks and how to disable those fallbacks on a host level and a network level.

FR4NK-W commented:
>I think we should shortly explain this in the README and maybe optionally also provide a command-line option to opt-out of this behavior.

Added the information about the potential use of the `whoami` and the public resolver fallback to the README.

I opted against adding a command-line option or configuration setting as it does not seem worth the effort to make the fallback configurable (How did we get there in the first place, if things are properly configured?), but instead opted to provide information on how to disable the fallbacks directly on the host or in network if really desired.

marcfrei commented on README.md:
```diff
@@ -13,6 +13,34 @@ It uses the following hinting mechanisms:
 - DNS-SD: DNS service discovery [RFC6763]
 - mDNS: multicast DNS [RFC6762]
 
+If the host being bootstrapper has no DNS search domain set, the rDNS functionality of DNS (as described in RFC1035)
```
"bootstrapper" > "bootstrapped"

FR4NK-W replied:
Done.

FR4NK-W commented on README.md:
```diff
@@ -13,6 +13,34 @@ It uses the following hinting mechanisms:
+
+All this is only a further fallback to provide zero-configuration bootstrapping even in misconfigured networks.
+
+Calls to these two third party services can be disabled entirely for a host by null routing their IP with the
```
Done.

FR4NK-W commented:
rebased onto master.
